File: scripts/durable-probe.mjs
// SPDX-License-Identifier: Apache-2.0
// Recovery-only synthetic host. Any prompt/provider/tool use is a test failure.
import {readFileSync} from 'node:fs';
import {WorkflowStore,OwnerCoordinator} from '@psp-cdl/api-server/persistence';
import {SqliteBackend} from '@psp-cdl/api-server/sqlite';
import {DurableLlmLoop,LoopError} from '@psp-cdl/llmproxy';
import {McpDispatchGate,bindingDigest} from '@psp-cdl/mcpproxy';
const options=JSON.parse(readFileSync(0,'utf8')),events=[];
const unexpected=()=>{throw Error('Unexpected inference/dispatch during recovery or lockdown');};
const host={authenticate:()=>({...options.actor,scopes:['sessions:read','sessions:write','models:invoke']}),now:()=>100,
  snapshot:(_p,s)=>({revision:'a1',policyVersion:s.policyVersion,registryRevision:'r1',providerId:'mock',providerRevision:'model-1',expires:190,releaseSources:[],releaseComplete:true}),
  prompt:unexpected,verification:unexpected,policy:unexpected,authorizeFinal:unexpected,planTurn:unexpected,authorizeTransition:unexpected,
  audit:(_p,event)=>{events.push(event);return true;},authorizeRecovery:()=>true,
  recoveryPolicy:(_p,b,r)=>({bindingDigest:bindingDigest(b),resources:[{classes:[],covenants:r.retained.covenants,capabilities:[],checks:{},parameters:{},context:{}}]})};
const backend=new SqliteBackend(process.argv[2],'peer-epoch',host.now);
try {
  const store=new WorkflowStore(backend,{resumeSecret:new Uint8Array(32).fill(42),authorizePersistence:()=>true,coordinator:new OwnerCoordinator(),durableTurns:true});
  const gate=new McpDispatchGate(store,host,'r1',[]);
  const loop=new DurableLlmLoop(store,gate,host,{id:'mock',revision:'model-1',complete:true,sources:[],invoke:unexpected},{postCompletion:'lockdown'});
  const recovered=await loop.recover('synthetic',options.sessionId,options.requestId,{deadline:180,cancelled:()=>false});
  let denied;
  try {await loop.run('synthetic',options.sessionId,{message:'Try to continue.'},{deadline:180,cancelled:()=>false,maxSteps:1,requestId:'new',expectedVersion:2});}
  catch(e){if(!(e instanceof LoopError))throw e;denied={code:e.code,response:e.response};}
  console.log(JSON.stringify({recovered,denied,events}));
}finally{backend.close();}

File: scripts/durable_probe.py
# SPDX-License-Identifier: Apache-2.0
"""Recovery-only synthetic host. Prompt/provider/tool use is a test failure."""
import json
import sys
from psp_cdl_api_server.persistence import WorkflowStore, OwnerCoordinator
from psp_cdl_api_server.sqlite import SqliteBackend
from psp_cdl_llmproxy import DurableLlmLoop, LoopError
from psp_cdl_mcpproxy import McpDispatchGate, binding_digest

sys.stdin.reconfigure(encoding="utf-8")
sys.stdout.reconfigure(encoding="utf-8")
options=json.load(sys.stdin)
events=[]
def unexpected(*_): raise AssertionError("Unexpected inference/dispatch during recovery or lockdown")
class Host:
    def authenticate(self,_): return {**options["actor"],"scopes":["sessions:read","sessions:write","models:invoke"]}
    def now(self): return 100
    def snapshot(self,p,s): return {"revision":"a1","policyVersion":s["policyVersion"],"registryRevision":"r1","providerId":"mock","providerRevision":"model-1","expires":190,"releaseSources":[],"releaseComplete":True}
    prompt=verification=policy=authorize_final=plan_turn=authorize_transition=unexpected
    def audit(self,p,event):
        events.append(event)
        return True
    def authorize_recovery(self,*_): return True
    def recovery_policy(self,p,b,r): return {"bindingDigest":binding_digest(b),"resources":[{"classes":[],"covenants":r["retained"]["covenants"],"capabilities":[],"checks":{},"parameters":{},"context":{}}]}
host=Host()
backend=SqliteBackend(sys.argv[1],"peer-epoch",host.now)
try:
    store=WorkflowStore(backend,resume_secret=bytes([42])*32,authorize_persistence=lambda *_:True,coordinator=OwnerCoordinator(),durable_turns=True)
    gate=McpDispatchGate(store,host,"r1",[])
    loop=DurableLlmLoop(store,gate,host,{"id":"mock","revision":"model-1","complete":True,"sources":[],"invoke":unexpected},{"postCompletion":"lockdown"})
    recovered=loop.recover("synthetic",options["sessionId"],options["requestId"],{"deadline":180,"cancelled":lambda:False})
    denied=None
    try: loop.run("synthetic",options["sessionId"],{"message":"Try to continue."},{"deadline":180,"cancelled":lambda:False,"maxSteps":1,"requestId":"new","expectedVersion":2})
    except LoopError as exc: denied={"code":exc.code,"response":getattr(exc,"response",None)}
    print(json.dumps({"recovered":recovered,"denied":denied,"events":events},ensure_ascii=False))
finally: backend.close()
